### PR TITLE
MESH-1220: Refactor & Optimize compliance manager

### PR DESCRIPTION
**Pull request by Centril**
Notable changes:
- `fn ensure_can_modify_rules` is introduced to refactor away copied logic.
- A `.dedup()` call was missing, now added.
- `fn verify_compliance_complexity` gets a simpler interface.
- Adding & removing default trusted issuers now requires less work in general and more expensive work (e.g., verifying complexity) moved to be later checks.
- `fn fetch_context`, `fn fetch_claims`, and `fn are_all_conditions_satisfied` is rendered as lazy as is possible, avoiding additional allocations, etc. Moreover, cloning the trusted issuers of a condition is also nixed by storing the defaults in the parent's stack, which also allows caching across all conditions in `fn are_all_conditions_satisfied` while fetching default trusted issuers only when actually needed.

It's probably easiest to review commit-by-commit.

### CODE REVIEW

**Review by CJP10 — APPROVED** [review]

File: pallets/compliance-manager/src/lib.rs [hunk 1]
```diff
@@ -574,45 +577,70 @@ impl<T: Trait> Module<T> {
...
-            condition.issuers.clone()
-        } else {
-            Self::trusted_claim_issuer(ticker)
-        };
+    ) -> proposition::Context<impl 'a + Iterator<Item = Claim>> {
+        enum Iter<I1, I2, I3> {
+            I1(I1),
+            I2(I2),
+            I3(I3),
+            I4,
```
Comment (maxsam4) [inline pallets/compliance-manager/src/lib.rs:615]: Some comments should be added to provide clarity.

**Review by maxsam4 — APPROVED** [review]
LGTM overall

**Review by maxsam4 — APPROVED** [review]

File: pallets/compliance-manager/src/lib.rs [hunk 2]
```diff
@@ -562,68 +536,96 @@ decl_event!(
...
 
 impl<T: Trait> Module<T> {
+    /// Ensure that `origin` can modify `ticker`'s compliance rules.
+    fn ensure_can_modify_rules(
+        origin: T::Origin,
+        ticker: Ticker,
+    ) -> Result<IdentityId, DispatchError> {
+        let sender = ensure_signed(origin)?;
+        CallPermissions::<T>::ensure_call_permissions(&sender)?;
+        let did = Context::current_identity_or::<Identity<T>>(&sender)?;
```
Comment (vkomenda) [inline pallets/compliance-manager/src/lib.rs:546]:
```suggestion
        let did = Identity::<T>::ensure_origin_call_permissions(&sender)?.primary_did;
```
We may not need to import `ensure_origin` or `Context`, or define `CallPermissions`.

File: pallets/compliance-manager/src/lib.rs [hunk 3]
```diff
@@ -562,68 +536,96 @@ decl_event!(
...
+        if condition.issuers.is_empty() {
+            slot.get_or_insert_with(|| Self::trusted_claim_issuer(ticker))
+        } else {
+            &condition.issuers
+        }
+    }
+
+    /// Fetches the proposition context for target `id` and specific `condition`.
+    /// Default trusted issuers, if fetched, are cached in `slot`.
+    fn fetch_context<'a>(
```
Comment (vkomenda) [inline pallets/compliance-manager/src/lib.rs:604]: Maybe a few words on the meaning of `Left` and `Right` in the return type?

**Review by vkomenda — APPROVED** [review]

**Review by satyamakgec — APPROVED** [review]

**Review by adamdossa — APPROVED** [review]

**Pull request merged**